hibernate-core/src/main/java/org/hibernate/boot/beanvalidation/TypeSafeActivator.java

```diff
@@ -263,14 +275,40 @@ private static boolean applyConstraints(
 					descriptor.getComposingConstraints(),
 					property, propertyDesc, null,
 					canApplyNotNull,
+					isConstraintCompositionOfTypeOr(descriptor),
 					dialect
 			);
 
-			hasNotNull = hasNotNull || hasNotNullFromComposingConstraints;
+			hasNotNull = compositionLogic.apply( hasNotNull, hasNotNullFromComposingConstraints );
+		}
+		if ( hasNotNull ) {
+			markNotNull( property );
 		}
 		return hasNotNull;
 	}
 
+	private static boolean isConstraintCompositionOfTypeOr(ConstraintDescriptor<?> descriptor) {
+		if ( descriptor.getComposingConstraints() == null ||
+				descriptor.getComposingConstraints().isEmpty() ) {
+			return false;
+		}
+
+		// This check assumes that Hibernate Validator is being used providing an
+		// implementation that gives us the composition type that is being used.
+		try {
+			Method compositionTypeMethod = descriptor.getClass()
```

Comment:
This looks like it's being executed many times.

Reply:
I think I can shortcut it a bit more. Should only need to trigger when there is more than one constraint in the composition.

Reply:
I think it should be cached statically.

Reply:
Ran code coverage on it based on the tests.
For the early-return part this gives 76 hits, 71 return early, 5 fall through.
Let met know if you consider this acceptable. Otherwise some kind of caching needs to be applied.

Reply:
Added static caching.